query-engine/connectors/sql-query-connector/src/ser_raw.rs

```diff
@@ -0,0 +1,553 @@
+use quaint::{
+    connector::{ResultRowRef, ResultSet},
+    Value, ValueType,
+};
+use serde::{ser::*, Serialize, Serializer};
+
+pub struct SerializedResultSet(pub ResultSet);
+
+#[derive(Debug, Serialize)]
+struct InnerSerializedResultSet<'a> {
+    columns: SerializedColumns<'a>,
+    types: &'a SerializedTypes,
+    rows: SerializedRows<'a>,
+}
+
+impl serde::Serialize for SerializedResultSet {
+    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let this = &self.0;
+
+        InnerSerializedResultSet {
+            columns: SerializedColumns(this),
+            types: &SerializedTypes::new(this),
+            rows: SerializedRows(this),
+        }
+        .serialize(serializer)
+    }
+}
+
+#[derive(Debug)]
+struct SerializedColumns<'a>(&'a ResultSet);
+
+impl<'a> Serialize for SerializedColumns<'a> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let this = &self.0;
+
+        if this.is_empty() {
+            return this.columns().serialize(serializer);
+        }
+
+        let first_row = this.first().unwrap();
+
+        let mut seq = serializer.serialize_seq(Some(first_row.len()))?;
+
+        for (idx, _) in first_row.iter().enumerate() {
+            if let Some(column_name) = this.columns().get(idx) {
+                seq.serialize_element(column_name)?;
+            } else {
+                // `query_raw` does not return column names in `ResultSet` when a call to a stored procedure is done
+                // See https://github.com/prisma/prisma/issues/6173
+                seq.serialize_element(&format!("f{idx}"))?;
+            }
+        }
+
+        seq.end()
+    }
+}
+
+#[derive(Debug, Serialize)]
+#[serde(transparent)]
+struct SerializedTypes(Vec<SerializedValueType>);
+
+impl SerializedTypes {
+    fn new(rows: &ResultSet) -> Self {
+        if rows.is_empty() {
+            return Self(Vec::with_capacity(0));
+        }
+
+        let row_len = rows.first().unwrap().len();
+        let mut types = vec![SerializedValueType::Unknown; row_len];
+        let mut types_found = 0;
+
+        'outer: for row in rows.iter() {
+            for (idx, value) in row.iter().enumerate() {
+                let current_type = types[idx];
+
+                if matches!(
+                    current_type,
+                    SerializedValueType::Unknown | SerializedValueType::UnknownArray
+                ) {
+                    let inferred_type = SerializedValueType::infer_from(value);
+
+                    if inferred_type != SerializedValueType::Unknown && inferred_type != current_type {
```

Comment:
Let's add a panic here in case `inferred_type != current_type` AND the `current_type` is not `Unknown`. If this bug ever comes into play, it will be very hard to debug without this check.

Reply:
I don’t understand the rationale. I think you’re trying to avoid a case where we update types that were already inferred, but it’s prevented by the if just above, isn't it?